src/main/java/com/greenfoxacademy/goddesstribesbackend/services/SoldierService.java

```diff
@@ -29,7 +30,11 @@ public int calculateFoodConsumptionRate(Long kingdomId) {
     ArrayList<Soldier> soldiers = findSoldiersByKingdom(kingdomId);
 
     for (Soldier soldier : soldiers) {
-      consumptionRate += soldier.getConsumptionRate();
+      int soldierConsumptionRate = soldier.getConsumptionRate();
+      if (LocalDateTime.now().isBefore(soldier.getFinishedAt())) {
+        soldierConsumptionRate -= Soldier.CONSUMPTION_RATE_PER_LEVEL;
+      }
+      consumptionRate += soldierConsumptionRate;
```

Comment:
Should not the `consumptionRate` be a computed in the service layer? It seems like you are substracting a number that you already added, going around the core problem: the data object is storing something that is transiently dependent to another value of it: the current level and whether it's updating or not, no?

Reply:
Should not the consumptionRate be a computed in the service layer? De igen, most ezt a SoldierService layerben számolom ki.

Amikor megszületik vagy upgradelődik egy farm/mine/townhall/barrack/soldier, akkor rögtön azzal a level-lel/generationRate-tel/consumptionRate-tel stb jön létre, amivel majd rendelkezni fog, amikor majd elkészül/upgradelődik. Ezért beállítjuk a finishedAt-et (ami egy későbbi időpont), és amíg nem érjük el ezt az időpontot (now < finishedAt), addig a level-je igazából level-1, a generationRate-je is az előző szinté, a consumptionRate-je is az előző szintével egyenlő. (De az adatbázisba már persze azzal a level-lel, generationRate-tel stb kerül be, amit majd el fog érni, ha elkészül)
Tehát, amíg pl a farm nem születik meg, addig nem termel kaját, és amíg nem upgradelődik fel, addig az előző szinten termel (tudod, Paks2, te mondtad...), és a katona is az előző szinten fogyasztja a kaját (itt ebben a kódban ezt látod).

Jobb lenne ezt szóban elmondani, így elég nehéz...

Reply:
Igen, egyébként a generationRate, consumptionRate mind függnek a level-től, szóval valóban kicsit fura ezeket példányadatként tárolni, hiszen elég lenne csak a level-t (a többi számolható ebből).